File: apps/android-native/app/src/main/java/com/alarmtalk/app/network/ApiErrorMessages.kt
package com.alarmtalk.app.network

import android.content.Context
import androidx.annotation.StringRes
import com.alarmtalk.app.R

/**
 * **에러 코드 → 사용자에게 보여 줄 문구.**
 *
 * 서버는 4xx/5xx 에서 `error_code` 를 함께 내려준다(목록은
 * `packages/shared/src/schemas/error-codes.ts`). 그 코드로 문구를 고르는 이유는 두 가지다:
 *
 * 1. **서버 문장은 우리 말이 아니다.** 상당수가 영어이고("Voice profile not found"),
 *    번역도 없다. 예전에는 [userFacingError] 가 "한글이 섞여 있으면 그대로 보여 준다" 는
 *    휴리스틱으로 걸러 냈는데, 그건 서버가 한국어를 쓰는 동안만 맞는 이야기다.
 * 2. **같은 실패에 늘 같은 말을 하려면** 문구가 화면이 아니라 코드에 붙어야 한다.
 *
 * ⚠ **화면별 문구가 우선이다.** 여기 표는 **마지막에서 두 번째** 층이다 — 화면이 자기
 * 맥락에 맞는 말을 갖고 있으면(목소리 등록 중의 `VOICE_FEATURE_REQUIRES_PAID_PLAN` 은
 * "유료 이용권에서 쓸 수 있어요") 그쪽이 이긴다. 이 표는 **아무도 안 맡은 코드**를 받는다.
 *
 * 층은 셋이다: 화면별 `when` → 이 표 → [userFacingError] 의 폴백 문장.
 *
 * ⚠ **모든 코드에 문구를 둘 필요는 없다.** `INVALID_JSON` 처럼 사용자가 할 수 있는 게
 * 없는 것은 비워 두고 폴백에 맡긴다 — 억지로 채우면 "요청 형식이 올바르지 않습니다" 같은
 * 알아들을 수 없는 말이 늘어난다.
 */
private val API_ERROR_MESSAGES: Map<String, Int> = mapOf(
    // ── 공통 ──────────────────────────────────────────────────────────────
    "RATE_LIMITED" to R.string.api_error_rate_limited,
    "REQUEST_BODY_TOO_LARGE" to R.string.api_error_request_too_large,
    "INTERNAL_ERROR" to R.string.api_error_server,
    "ALARM_SCHEMA_UPGRADING" to R.string.api_error_schema_upgrading,

    // ── 로그인·계정 ───────────────────────────────────────────────────────
    "AUTH_INVALID_CREDENTIALS" to R.string.auth_error_invalid_credentials,
    "AUTH_EMAIL_TAKEN" to R.string.msg_register_email_taken,
    "AUTH_EMAIL_SOCIAL" to R.string.msg_register_email_social_google,
    "AUTH_EMAIL_CODE_INVALID" to R.string.api_error_email_code_invalid,
    "AUTH_EMAIL_CODE_EXPIRED" to R.string.api_error_email_code_expired,
    "AUTH_EMAIL_CODE_ATTEMPTS_EXCEEDED" to R.string.api_error_email_code_attempts,
    "ACCOUNT_PENDING_DELETION" to R.string.api_error_account_pending_deletion,
    // 세션이 끊긴 셋. 앱은 이미 자동 로그아웃으로 처리하지만, 그 사이 스낵바가 뜨면
    // "다시 로그인" 이라고 말해야 사용자가 무슨 일인지 안다.
    "TOKEN_REVOKED" to R.string.api_error_session_expired,
    "AUTH_TOKEN_EXPIRED" to R.string.api_error_session_expired,
    "AUTH_USER_NOT_FOUND" to R.string.api_error_session_expired,

    // ── 이용권 게이트 ─────────────────────────────────────────────────────
    "VOICE_FEATURE_REQUIRES_PAID_PLAN" to R.string.plan_gate_paid_message,
    "VOICE_LOCKED_FREE_PLAN" to R.string.msg_voice_locked_free_plan,
    "FREE_PLAN_PRESET_ONLY" to R.string.msg_voice_preset_only,
    "BASIC_VOICE_PRESET_ONLY" to R.string.msg_voice_preset_only,
    "MANUAL_TTS_QUOTA_EXCEEDED" to R.string.editor_error_manual_tts_quota,

    // ── 목소리 ────────────────────────────────────────────────────────────
    "VOICE_CLONE_AUDIO_TOO_SHORT" to R.string.msg_voice_clone_audio_too_short,
    "VOICE_CLONE_AUDIO_TOO_LONG" to R.string.msg_voice_clone_audio_too_long,
    "INVALID_DURATION" to R.string.msg_voice_invalid_duration,
    "INVALID_AUDIO_MIME_TYPE" to R.string.msg_voice_invalid_audio_format,
    "VOICE_SLOT_EXHAUSTED" to R.string.msg_voice_slot_exhausted,
    "VOICE_CAPACITY_EXHAUSTED" to R.string.msg_voice_slot_exhausted,
    "VOICE_LIMIT_REACHED" to R.string.api_error_voice_limit_reached,
    "VOICE_NOT_READY" to R.string.api_error_voice_not_ready,
    "VOICE_CLONING_FAILED" to R.string.api_error_voice_cloning_failed,
    "VOICE_MONTHLY_CHANGE_LIMIT_REACHED" to R.string.msg_voice_monthly_change_limit,
    "VOICE_PREVIEW_REQUIRED" to R.string.msg_voice_preview_required,
    "CONSENT_REQUIRED" to R.string.msg_voice_consent_required,

    // ── 알람·문구 ─────────────────────────────────────────────────────────
    "ALARM_NOT_FOUND" to R.string.api_error_alarm_not_found,
    "TTS_GENERATION_FAILED" to R.string.api_error_tts_generation_failed,
)

/** 이 코드에 정해 둔 문구가 있으면 돌려준다. 없으면 null — 부르는 쪽이 폴백을 쓴다. */
@StringRes
fun apiErrorMessageRes(code: String?): Int? = code?.let { API_ERROR_MESSAGES[it] }

/**
 * 코드에 정해 둔 문구. 없으면 null.
 *
 * ⚠ **폴백을 여기서 만들지 않는다.** 서버 문장을 그대로 쓸지 화면이 준 문장을 쓸지는
 * `ui/util/PlatformAndLabelUtils.kt` 의 `userFacingError` 가 정한다 — 규칙이 두 곳에
 * 생기면 한쪽만 고쳐진다. 그래서 호출부는 늘 이렇게 잇는다:
 *
 * ```
 * message = apiErrorMessage(app, code) ?: userFacingError(error, app.getString(R.string.…))
 * ```
 *
 * ⚠ **`code` 는 이미 뽑아 놓은 값을 넘길 것.** `HttpException` 의 errorBody 는 **한 번만**
 * 읽힌다 — 같은 예외에 [apiErrorCode] 를 두 번 부르면 두 번째가 빈손이 된다.
 */
fun apiErrorMessage(context: Context, code: String?): String? =
    apiErrorMessageRes(code)?.let(context::getString)

File: apps/ios-native/AlarmTalk/APIErrorMessages.swift
import Foundation

/// **에러 코드 → 사용자에게 보여 줄 문구.**
///
/// 서버는 4xx/5xx 에서 `error_code` 를 함께 내려준다(목록은
/// `packages/shared/src/schemas/error-codes.ts`). 그 코드로 문구를 고르는 이유는 두 가지다:
///
/// 1. **서버 문장은 우리 말이 아니다.** 상당수가 영어이고("Voice profile not found") 번역도
///    없다. 예전에는 `containsKorean` 휴리스틱으로 걸러 냈는데, 그건 서버가 한국어를 쓰는
///    동안만 맞는 이야기다.
/// 2. **같은 실패에 늘 같은 말을 하려면** 문구가 화면이 아니라 코드에 붙어야 한다.
///
/// ⚠ **화면별 문구가 우선이다.** 이 표는 **마지막에서 두 번째** 층이다 — 목소리 화면처럼
/// 자기 맥락에 맞는 말을 가진 곳(`VoiceStudioViewModel+ErrorMapping`)이 이기고, 여기는
/// **아무도 안 맡은 코드**를 받는다. 층은 셋: 화면별 매핑 → 이 표 → 일반 폴백 문장.
///
/// ⚠ **안드로이드 `network/ApiErrorMessages.kt` 와 같은 표다.** 한쪽에만 코드를 더하면
/// 같은 실패가 두 앱에서 다르게 읽힌다 — 「iOS 는 안드로이드를 원본으로 삼는다」.
enum APIErrorMessages {
    /// 이 코드에 정해 둔 문구. 없으면 nil — 부르는 쪽이 자기 폴백을 쓴다.
    ///
    /// ⚠ **모든 코드에 문구를 둘 필요는 없다.** `INVALID_JSON` 처럼 사용자가 할 수 있는 게
    /// 없는 것은 비워 두고 폴백에 맡긴다 — 억지로 채우면 알아들을 수 없는 말이 늘어난다.
    static func message(for code: String?) -> String? {
        guard let code else { return nil }
        switch code {
        // ── 공통 ──────────────────────────────────────────────────────────
        case "RATE_LIMITED":
            return String(localized: "요청이 너무 많아요. 잠시 후 다시 시도해 주세요.")
        case "REQUEST_BODY_TOO_LARGE":
            return String(localized: "보내려는 파일이 너무 커요. 더 짧게 녹음해 주세요.")
        case "INTERNAL_ERROR":
            return String(localized: "잠시 문제가 생겼어요. 잠시 후 다시 시도해 주세요.")
        case "ALARM_SCHEMA_UPGRADING":
            return String(localized: "서버를 업데이트하고 있어요. 잠시 후 다시 시도해 주세요.")

        // ── 로그인·계정 ───────────────────────────────────────────────────
        case "AUTH_INVALID_CREDENTIALS":
            return String(localized: "이메일 또는 비밀번호를 확인해 주세요.")
        case "AUTH_EMAIL_CODE_INVALID":
            return String(localized: "인증 코드가 맞지 않아요. 다시 확인해 주세요.")
        case "AUTH_EMAIL_CODE_EXPIRED":
            return String(localized: "인증 코드가 만료됐어요. 다시 받아 주세요.")
        case "AUTH_EMAIL_CODE_ATTEMPTS_EXCEEDED":
            return String(localized: "인증 코드를 너무 많이 틀렸어요. 다시 받아 주세요.")
        case "ACCOUNT_PENDING_DELETION":
            return String(localized: "탈퇴 신청 중인 계정이에요. 탈퇴를 취소하면 다시 쓸 수 있어요.")
        // 세션이 끊긴 셋. 앱은 자동 로그아웃으로도 처리하지만, 그 사이 문구가 뜨면
        // "다시 로그인" 이라고 말해야 사용자가 무슨 일인지 안다.
        case "TOKEN_REVOKED", "AUTH_TOKEN_EXPIRED", "AUTH_USER_NOT_FOUND":
            return String(localized: "로그인이 만료됐어요. 다시 로그인해 주세요.")

        // ── 이용권 게이트 ─────────────────────────────────────────────────
        case "VOICE_FEATURE_REQUIRES_PAID_PLAN":
            return PaidGateCopy.message
        case "VOICE_LOCKED_FREE_PLAN":
            return String(localized: "무료 이용권으로 바뀌어 이 목소리를 쓸 수 없어요. 이용권을 다시 등록하면 돌아와요.")
        case "FREE_PLAN_PRESET_ONLY", "BASIC_VOICE_PRESET_ONLY":
            return String(localized: "기본 목소리는 준비된 문구로만 말할 수 있어요. 직접 입력한 문구로 깨우려면 내 목소리를 골라 주세요.")
        case "MANUAL_TTS_QUOTA_EXCEEDED":
            return String(localized: "이번 달 직접 입력 문구 만들기 횟수를 다 썼어요. 다음 달에 다시 채워져요.")

        // ── 목소리 ────────────────────────────────────────────────────────
        case "VOICE_CLONE_AUDIO_TOO_SHORT":
            return String(localized: "목소리를 만들 음성은 12초 이상이어야 해요.")
        case "VOICE_CLONE_AUDIO_TOO_LONG":
            return String(localized: "목소리를 만들 음성은 2분 이하로 준비해 주세요.")
        case "INVALID_DURATION":
            return String(localized: "음성 길이를 확인하지 못했어요. 파일을 다시 선택해 주세요.")
        case "INVALID_AUDIO_MIME_TYPE":
            return String(localized: "지원하지 않는 오디오 형식이에요. 다시 녹음해 주세요.")
        case "VOICE_SLOT_EXHAUSTED", "VOICE_CAPACITY_EXHAUSTED":
            return String(localized: "지금은 목소리 생성 요청이 많아요. 잠시 후 다시 시도해 주세요.")
        case "VOICE_LIMIT_REACHED":
            return String(localized: "등록할 수 있는 목소리를 다 채웠어요. 쓰지 않는 목소리를 지워 주세요.")
        case "VOICE_NOT_READY":
            return String(localized: "목소리를 아직 만들고 있어요. 잠시 후 다시 시도해 주세요.")
        case "VOICE_CLONING_FAILED":
            return String(localized: "목소리를 만들지 못했어요. 조용한 곳에서 다시 녹음해 주세요.")
        case "VOICE_MONTHLY_CHANGE_LIMIT_REACHED":
            return String(localized: "목소리는 한 달에 1번만 바꿀 수 있어요. 다음 달에 다시 시도해 주세요.")
        case "VOICE_PREVIEW_REQUIRED":
            return String(localized: "문구가 바뀌었어요. 새 문구를 끝까지 들어본 뒤 저장해 주세요.")
        case "CONSENT_REQUIRED":
            return String(localized: "목소리를 만들려면 음성 정보 활용 동의가 필요해요. 더보기 → 약관 및 동의에서 다시 동의해 주세요.")

        // ── 알람·문구 ─────────────────────────────────────────────────────
        case "ALARM_NOT_FOUND":
            return String(localized: "이미 사라진 알람이에요.")
        case "TTS_GENERATION_FAILED":
            return String(localized: "음성을 만들지 못했어요. 잠시 후 다시 시도해 주세요.")

        default:
            return nil
        }
    }
}
